slick/src/main/scala/slick/util/ManagedArrayBlockingQueue.scala

```diff
@@ -99,169 +83,121 @@ abstract class ManagedArrayBlockingQueue[E >: Null <: AnyRef](capacity: Int, fai
-    if (c eq this) throw new IllegalArgumentException
-    val items = this.items
+  def drainTo(c: util.Collection[_ >: E]): Int = {
     locked {
-      var i = takeIndex
-      var n = 0
-      val max = count
-      while (n < max) {
-        c.add(items(i).asInstanceOf[E])
-        items(i) = null
-        i = inc(i)
-        n += 1
-      }
+      val n = highPrioItemQueue.drainTo(c) + itemQueue.drainTo(c)
       if (n > 0) {
-        count = 0
-        putIndex = 0
-        takeIndex = 0
-        notFull.signalAll
+        notFull.signalAll()
       }
       n
     }
   }
 
-  def drainTo(c: Collection[_ >: E], maxElements: Int): Int = {
-    checkNotNull(c)
-    if (c eq this) throw new IllegalArgumentException
-    if (maxElements <= 0) return 0
-    val items = this.items
+  def drainTo(c: util.Collection[_ >: E], maxElements: Int): Int = {
     locked {
-      var i: Int = takeIndex
-      var n: Int = 0
-      val max: Int = if ((maxElements < count)) maxElements else count
-      while (n < max) {
-        c.add(items(i).asInstanceOf[E])
-        items(i) = null
-        i = inc(i)
-        n += 1
+      var n = highPrioItemQueue.drainTo(c, maxElements)
+      if (n < maxElements) {
+        n += itemQueue.drainTo(c, maxElements - n)
       }
       if (n > 0) {
-        count -= n
-        takeIndex = i
-        notFull.signalAll
+        notFull.signalAll()
       }
       n
     }
   }
 
-  def iterator: Iterator[E] = new Iterator[E] {
-    private var remaining: Int = _
-    private var nextIndex: Int = _
-    private var nextItem: E = _
-    private var lastItem: E = _
-    private var lastRet: Int = -1
+  def iterator: util.Iterator[E] = new util.Iterator[E] {
 
-    locked {
-      remaining = count
-      if(remaining > 0) {
-        nextIndex = takeIndex
-        nextItem = itemAt(nextIndex)
-      }
-    }
+    private var current = 0
+    private val iterators = Array(highPrioItemQueue.iterator, itemQueue.iterator)
 
-    def hasNext: Boolean = remaining > 0
+    override def hasNext: Boolean = {
+      while (current < iterators.length && !iterators(current).hasNext)
+        current = current + 1
+
+      current < iterators.length
+    }
 
     def next: E = {
       locked {
-        if (remaining <= 0) throw new NoSuchElementException
-        lastRet = nextIndex
-        var x: E = itemAt(nextIndex)
-        if (x == null) {
-          x = nextItem
-          lastItem = null
-        }
-        else lastItem = x
-        while ({ remaining -= 1; remaining > 0 } && { nextIndex = inc(nextIndex); nextItem = itemAt(nextIndex); nextItem == null }) ()
-        x
+        while (current < iterators.length && !iterators(current).hasNext)
+          current = current + 1
+
+        return iterators(current).next()
       }
     }
 
-    def remove: Unit = throw new UnsupportedOperationException
+    def remove(): Unit = throw new UnsupportedOperationException
   }
 
   @inline private[this] def locked[T](f: => T) = {
```

Comment:
This will only ever have a chance at being inlined if -optimize is used when compiling